Add search blog

coding-in-public opened the pull request:
Okay, couple of things here:
- I created the fetched endpoint as a dynamic endpoint, but I'm wondering if I should have just made it a single fetch to the blog collection and then do all the fuseJS stuff on the client? It seems slow? But it's quite possible I missed something
- I’ve prepped the searchResults component to be able to progressively reveal more posts (either by clicking a button or by creating an infinite scroll)

I know this may take a while to review and I'm more than happy to adjust anything you see. I'm sure I've made some TypeScript faux pas, too, so feel free to point me in the right direction there. Hope this is at least close to what you were thinking?

jsjoeio commented:
I don't know if you can see the build logs but here is what it says:

```shell
error   Named export 'CgSpinner' not found. The requested module 'react-icons/cg/index.esm.js' is a CommonJS module, which may not support all module.exports as named exports.
--
21:58:54.370 | CommonJS modules can always be imported via the default export, for example using:
21:58:54.370 |  
21:58:54.370 | import pkg from 'react-icons/cg/index.esm.js';
21:58:54.370 | const { CgSpinner } = pkg;
```

jsjoeio commented:
Awesome! I'm signing off for the night but will look at this tomorrow when I start my work day. Excited :D

coding-in-public commented:
Ah, yeah, weird. I just removed the react-icons package. Might as well keep the dependencies down anyhow. Substituted them for inline SVGs and then also removed the auto-focus on the blog page.

coding-in-public commented:
Also, I sent this through right before I went to bed last night and looking at it this AM, it’s rendering oddly. For whatever reason, in local it's only updating the search results when I open/close the dev tools. I don't think I'll get time to work on this today (maybe tomorrow)? But I'm wondering if either
- I have a bug? 
OR
- the SSR .json endpoint is too slow (I'm guessing that's re-rendering with each keystroke?)

My plan would be to move the FuseJS processing and everything to the client and only pull that JSON file on initial page load once. Then all the rest of the rendering would be client side. Eventually, this may cause performance problems, but right now the JSON file is SUPER small and will be for some time until you add a lot of other posts. 

If you have time or if you have another idea in the meantime, go for it. :)

jsjoeio commented:
> My plan would be to move the FuseJS processing and everything to the client and only pull that JSON file on initial page load once. Then all the rest of the rendering would be client side.

I like this a lot! If My preference is to go with this approach because it's straightforward then solve the performance issue _if/when_ it happens.

> If you have time or if you have another idea in the meantime, go for it. :)

Cool! May get to it tomorrow or this weekend if you don't first

coding-in-public commented:
Hey, had 15 min tonight and I think I got everything moved over. There are a couple of TS errors that I'm sure are basic…but didn't have time to chase them down. I _think_ everything else is working! Let me know what you think.

jsjoeio commented on src/components/PostCard.tsx:
```diff
@@ -0,0 +1,35 @@
+          dateTime={new Date(post.data.date).toISOString()}
+          className="uppercase text-arg-gold"
+        >
+          {new Date(post.data.date).toLocaleDateString("en-us", {
```
Not sure if case-sensitive: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString#using_locales

```suggestion
          {new Date(post.data.date).toLocaleDateString("en-US", {
```

jsjoeio commented on src/components/PostCard.tsx:
```diff
@@ -0,0 +1,35 @@
+        <h2 className="text-xl sm:text-2xl max-w-heading font-bold">
+          <a
+            className="font-bold focus:outline-none focus-visible:ring-4 ring-offset-2 ring-offset-arg-white ring-arg-gold"
+            href={`/${post.slug}/`}
```
Do we want/need the trailing slash here? My guess is no? 
```suggestion
            href={`/${post.slug}`}
```

coding-in-public replied:
I think it creates a redirect if you don’t do the trailing slash since Astro puts it in by default (if I'm remembering correctly); might be worth testing

jsjoeio commented on src/components/PostCard.tsx:
```diff
@@ -0,0 +1,35 @@
+      <div className="bg-arg-muted h-[1px] w-full" />
+      <p>{post.data.description}</p>
+      <a
+        href={`/${post.slug}/`}
```
Same as above:
```suggestion
        href={`/${post.slug}`}
```

jsjoeio commented on src/components/search/SearchApp.tsx:
```diff
@@ -0,0 +1,39 @@
+// component imports
+import SearchForm from "./SearchForm";
+import SearchResults from "./SearchResults";
+const queryClient = new QueryClient();
```
Does this need to be inside `SearchApp`? I haven't used `react-query` in a while

coding-in-public replied:
What do you mean by "this"? the query client? or the component imports? or both? 

I believe you have to wrap any components you want to access your useQuery hook with the queryClient provider?

jsjoeio replied:
Oops, I meant this: 

```ts
const queryClient = new QueryClient();
```

But I just checked the docs and see the example here: https://tanstack.com/query/v4/docs/react/adapters/react-query#example

TIL! Disregard this!

jsjoeio commented on src/components/search/SearchApp.tsx:
```diff
@@ -0,0 +1,39 @@
+    const url = new URL(window.location.href);
+    if (query) {
+      url.searchParams.set("q", query);
+      window.history.replaceState(null, "", url.toString());
```
❓ asking to learn- why do we need to do this `history.replaceState` thing here?

jsjoeio replied:
Ahhh got it! That makes sense!

jsjoeio commented on src/components/search/SearchResults.tsx:
```diff
@@ -0,0 +1,88 @@
+  score: number;
+};
+
+const DEFAULT_NUM = 5;
```
nit: could we get more specific with this variable name? Like `DEFAULT_NUM_RESULTS_SHOWN` or something like that?

coding-in-public replied:
Yeah, I like that!

jsjoeio commented on src/components/search/SearchResults.tsx:
```diff
@@ -0,0 +1,88 @@
+    },
+  });
+
+  const fuse = new Fuse(data, {
```
Super cool! I've never used this before but looks like a nice API 🔥

jsjoeio replied:
so mine is the guinea pig? 😂

coding-in-public replied:
lol, well if it means anything, I implemented this on another blog this week first. So you’re the second guinea pig? lol And then I'll do mine tomorrow probably for my blog. Just bundling over here

jsjoeio replied:
LOL mostly kidding but nice! Feeling good then!

jsjoeio commented on src/components/search/SearchResults.tsx:
```diff
@@ -0,0 +1,88 @@
+    }
+  }, [query, data, numOfResults]);
+
+  if (query.length === 0) return null;
```
nit: if no results, could we show a div saying like "no results found. try searching for something else?"

(although I could be misunderstanding how this works...)

jsjoeio replied:
Ahhh, I see now! Makes perfect sense 🙌

jsjoeio commented on src/components/ui/LoadingSpinner.jsx:
Can we make this a `.tsx` file? I bet ChatGPT could convert if you need help with the types

coding-in-public replied:
oh lol. Didn't even realize I hadn't. There aren't really any types to worry about, so I think we could just change it to a .tsx extension and it should work.

jsjoeio commented on src/pages/blog.astro:
```diff
@@ -29,32 +28,9 @@ const posts = (await getCollection("blog")).sort(
-        }
+        {posts.map((post) => <PostCard post={post} key={post.slug} />)}
       </section>
+      <SearchWidget />
```
Should we lazy load this with a client directive like `client:idle` or whatever you recommend?

coding-in-public replied:
I'd decided to make this widget an Astro component, so it'll just render as straight HTML (i.e., it's not a React component). It's basically a fancy redirect to the /search route with the React code that actually does all the searching.

jsjoeio replied:
ooooo I see now! Nice 🙌

jsjoeio commented on src/components/search/SearchWidget.astro:
```diff
@@ -0,0 +1,47 @@
+  </div>
+</aside>
+
+<script>
+  const searchWidget = document.querySelector(
+    "#search-widget"
+  ) as HTMLFormElement;
+
+  searchWidget.addEventListener("submit", (e) => {
+    e.preventDefault();
+    const formData = new FormData(searchWidget);
+    const searchInput = formData.get("search") as string;
+    const url = new URL("/search/", window.location.origin);
+    url.searchParams.set("q", searchInput);
+    window.location.href = url.toString();
+  });
+</script>
```
❓ I'm still new to this part of Astro. Why do we have to do this in a script tag rather than above in the frontmatter section?

jsjoeio replied:
Totally! Sounds great!

jsjoeio commented:
Also there is a bit of some content shift when you enter a query vs no query. I wonder how we can fix that 🤔 Any ideas?

https://user-images.githubusercontent.com/3806031/236255831-fac6ca25-ae89-4d55-8b22-a096de8f1bb7.mov

coding-in-public commented:
> Also there is a bit of some content shift when you enter a query vs no query. I wonder how we can fix that 🤔 Any ideas?
> 
>  Screen.Recording.2023-05-04.at.8.30.00.AM.mov

Ah, yeah. I think it's just the scrolling sidebar that pops in when the content overflows. I'm thinking we could do two things:
1. Have a permanent sidebar no matter what
- set overflow: scroll on the body
OR
- set the body section to be min-height 101vh
2. Do some fancy css to shift the content left as if there was a sidebar (when not present) and shift back to the right when there is a sidebar. It would be hard to get it consistent since the sidebar width is different in each browser. We could probably get the body’s client width or something with JS and calculate that movement?

jsjoeio commented:
> Ah, yeah. I think it's just the scrolling sidebar that pops in when the content overflows. I'm thinking we could do two things:

Ahh right. Okay I think that's safe to save for a follow-up task then. I'm hoping there's a simpler solution like either adding the CSS you suggested or adjusting the min-height or something

coding-in-public commented:
What's the next step here? (And this may be a dense question) but do you want me to make the suggested changes and re-push? or are you planning on making those changes and merging?

jsjoeio commented:
> What's the next step here? (And this may be a dense question) but do you want me to make the suggested changes and re-push

Ah good question! So my assumption is you:
- make suggested changes
- fix merge conflicts
- optional: make the vercel/astro changes for trailingSlash

Then re-request a review from me. Image below ICYDK:

<img width="360" alt="image" src="https://user-images.githubusercontent.com/3806031/236324761-82da3c0e-ed6c-4ecb-8347-89a4e8350641.png">

Then I'll review again and we can merge! 

I'll fix the layout shift in a separate PR (assuming you didn't want to, your call!)

coding-in-public commented:
I _think_ I did all that correctly? :)

jsjoeio commented:
Heck yeah! Looks good from the code side. Let me commit the `trailingSlash` changes to your branch real quick then merge!

jsjoeio reviewed (approved):
Wahoo!

coding-in-public commented:
https://user-images.githubusercontent.com/86967271/236368655-b568ef9c-6e2d-4432-a69e-218678ae32d5.mp4

🥳

jsjoeio commented:
@coding-in-public dude you killed it! looks so professional too 🚀
